### ISSUE TITLE
NNI 2021 Mar~Apr Iteration Planning

### ISSUE THREAD

**Issue by kvartet** [microsoft/nni#3445]
This is the plan for the iteration in Mar~Apr, it's a 4 weeks iteration.

# Release Plan for v2.2
- Release manager: @SparkSnail 
- Feature freeze date: 3.29
- Code freeze and demo date: ~4.6~ 4.9
- Release date: ~4.12~ ~4.19~ 4.25
-----
Let's try to adopt the legend of annotations:

| Mark | Description                                  |
| ---- | -------------------------------------------- |
| 🏃‍♂️   | work in progress                             |
| 👪    | work under review                            |
| ✋    | blocked task                                 |
| 💪    | stretch goal for this iteration              |
| 🔴    | missing issue / pull request reference       |
| 🔵    | more investigation and survey required       |
| ⚫    | under discussion within the team             |
| ⬛    | a large work item, larger than one iteration |

### nnictl & nni.experiment
- [ ] P0 - native config for v2 [tester:@98may] (testing)
- [x] P0 - make nni.experiment complete (view, resume) [tester:@kvartet] (need to update doc @J-Shang)
### Retiarii & NAS
- [x] P0 - format of output architecture: implementation and doc [tester: @colorjam]  
- [x] P0 - nested value choice [tester: @QuanluZhang @linbinskn ]
- [x] P0 - support dict value choice [tester: @QuanluZhang @linbinskn ]
- [x] P0 - Support debug mode (standalone mode) [tester: ~@ultmaster~ @linbinskn @colorjam ]
### Model compression
- [x] P0 - Quantization improvement: model import/export [tester: @colorjam ]
- [x] P0 - Quantization speedup tool [ tester: @colorjam]
- [x] P1 - Update TF model compression: support checkpoint (deferred from v2.0) [ tester: @J-Shang]
### Training service & NNI manager
- [x] P0 - Reuse mode envs dir iterate [tester:@acured]
- [x] P0 - Add log information for gpu_collector [tester:@liuzhe-lz]
- [x] P0 - Improve share storage [tester: @acured]
- [x] P1 - Reuse mode remote windows (set "reuse" as default) [tester: @SparkSnail]
- [x] P1 - Support tensorboard [tester: ~@ultmaster~ @colorjam ]

### WebUI
- [x] minor improvements: [ tester: @J-Shang]
    - [x] P0 - Upgrade echarts-for-react 
    - [x] P1 - webUI dispatcher log / nnimanager log wrap 
- [x] P1 - Support tensorboard [ tester: ~@ultmaster~ @colorjam ]
  - [x] UX design DL: 2.3
  - [x] backend
 
### Bug fix
- [x] Fix bug about hyper-parameter Add/Remove axes and table Add/Remove columns button conflict [tester: @SparkSnail]
- [x] Fix bug about monaco editor search text is not display completely [tester: @kvartet] 

### nnictl & nni.experiment
- [x] P0 - write v2 config files for examples @liuzhe-lz PR: #3388 (Depend on pr #3466, ready for review) 
     - [x] current items could be merged (fixing IT, 4/7)
- [ ] P0 - make nni.experiment complete (view, resume) @j-shang (@nihao backup) (,4/7 PR #3490 )

### Retiarii & NAS
- [x] P0 - format of output architecture: implementation and doc (need experiment api design meeting)@ultmaster (#3464 in review) 
- [x] P0 - nested value choice @ultmaster (#3508 need review) 
- [x] P0 - support dict value choice @ultmaster (#3508 need view) 
- [x] P0 - Support debug mode (standalone mode) @QuanluZhang 
- [x] P0 - Refactor NAS examples @QuanluZhang (PR #3513) 
- [x] P1 - AML & NAS integration demo @kvartet (4/2)
     - [ ] PR for notebook example and demo 

### Model compression
- [X] P0 - Quantization improvement: model import/export @linbinskn (3/19 PR #3458) 
- [x] P0 - Quantization speedup tool @linbinskn (#3488 ready review)
- [x] P1 - Update TF model compression: support checkpoint (deferred from v2.0) @liuzhe-lz (#3487 need review)
- [x] P1 - doc improvement @colorjam (PR #3482 need review) 

### Training service & NNI manager
- [x] P0 - Reuse mode envs dir iterate @SparkSnail (3/22 PR #3472)
- [x] P0 - Add log information for gpu_collector @SparkSnail (#3506, PR ready for review) 
- [ ] P0 - add test for shared storage @acured (3/24 #3455 need to fix IT and publish PR, depend on #3456) **ETA 4/6**
    - [x] P0 - Improve share storage @J-shang (3/19 #3456 need review)
- [x] P1 - Reuse mode remote windows (set "reuse" as default) (@acured ask @SparkSnail, 3/31 ) 
- [x] P1 - Support tensorboard @J-shang (#3454 need review) 

### WebUI
- [x] minor improvements: @Lijiaoa (3/19) 
    - [x] P0 - Upgrade echarts-for-react (#3457)
    - [x] P1 - webUI dispatcher log / nnimanager log wrap (#3461 need review)  @Lijiaoa 
- [x] P1 - Support tensorboard @Lijiaoa  (4/7 #3361)
  - [x] UX design DL: 2.3
  - [x] backend

### Pipelines
- [ ] P0 - add aml pipeline test @SparkSnail (#3477 draft pr) **ETA 4/7**
- [ ] P0 - add hybrid pipeline test @SparkSnail (#3478 draft pr) **ETA 4/7**
 
### Bug fix
- [x] Fix bug about hyper-parameter Add/Remove axes and table Add/Remove columns button conflict PR#3491 
- [x] Fix bug about monaco editor search text is not display completely issue #3484 PR: #3492

### Next release candidate
- [ ] P0 - TensorFlow support exploration @liuzhe-lz (does not follow feature freeze)
- [ ] P0 - More high-level APIs for users to easily express model space @ultmaster (4/9 #3481 will PR ready, may delay) **ETA 4/9**
- [ ] P1 - NAS visualization improvements: refactor graph representation (deferred from v2.0) @98may  
     - [ ] Step1. Data structure design @98may 
     - [ ] Step2. Visulization @98may 
- [ ] P1 - high level API design @J-shang (4/2 design, need discussion)
- [ ] P1 - customize dafault metric to multiple metrics (#3504 @kvartet )
- [ ] P1 - Bug: Tuner need to exit when experiment done
- [ ] P1 - module compression pipeline @colorjam  (4/2, will have followup task)
- [ ] P1 - Support a fancy NAS algorithm which could leverage architectural information @98may (PR #3479, need review)
- [ ] P0 - Detail Page improvement @Lijiaoa 
    - [ ] Refine search parameters #3024 #3099 
    - [ ] Put the add/remove button in the last column of the header #3099 (draft PR, #3485)
- [ ] P1 - [Retiarii] Further improve the logic of graph generation and code generation @QuanluZhang
  - [ ] Support control flow in graph/code generation
  - [ ] Further integrated with TorchScript
- [ ] P1 - Add pure-python execution engine @ultmaster (4/2)
- [ ] P1 - Support RL strategy @ultmaster (4/2)
- [ ] P1 - NAS visualization 
  - [ ] UX Design - concept design @QuanluZhang / @scarlett2018 
  - [ ] UX Design - formal design @yangou1988 
  - [ ] visualization implementation - @98may 
- [ ] P1 - set base url for the web report (#2771) @acured 
- [ ] P1 - Refactor readthedocs integration @ultmaster ()
- [ ] P0 - (discussion) support experiment resume/view/stop on webUI @Lijiaoa (depending with Quanlu meeting)
- [ ] P1 - Pruning model speedup improvement @J-shang
